src/operator/tensor/elemwise_binary_broadcast_op.h

```diff
@@ -0,0 +1,238 @@
+    SHAPE_ASSIGN_CHECK(*out_attrs, 0, lhs);
+    return true;
+  }
+  TShape out(std::max(lhs.ndim(), rhs.ndim()));
+  int bl = out.ndim() - lhs.ndim();
+  int br = out.ndim() - rhs.ndim();
+  for (int i = 0; i < out.ndim(); ++i) {
+    int l = 1, r = 1;
+    if (i >= bl) l = lhs[i-bl];
+    if (i >= br) r = rhs[i-br];
+    if (l != r) {
+      CHECK(l == 1 || r == 1) 
+        << "operands could not be broadcast together with shapes " << lhs << " " << rhs;
+      out[i] = std::max(l, r);
+    } else {
+      out[i] = l;
+    }
+  }
+  SHAPE_ASSIGN_CHECK(*out_attrs, 0, out);
+  return true;
+}
+
+bool BinaryBroadcastShapeCompact(const TShape& lshape, const TShape& rshape, const TShape& oshape,
+                                 TShape *new_lshape, TShape *new_rshape, TShape *new_oshape) {
+  if (lshape == rshape) return false;
+  *new_lshape = TShape(oshape.ndim());
+  *new_rshape = TShape(oshape.ndim());
+  *new_oshape = TShape(oshape.ndim());
+  int bl = oshape.ndim() - lshape.ndim();
+  int br = oshape.ndim() - rshape.ndim();
+  int j = 0, prod = 1;
+  for (int i = 0; i < oshape.ndim(); ++i) {
+    int l = 1, r = 1, o = oshape[i];
+    if (i >= bl) l = lshape[i-bl];
+    if (j >= br) r = rshape[i-br];
+    if (l != r) {
+      if (prod > 1) {
+        (*new_lshape)[j] = (*new_rshape)[j] = (*new_oshape)[j] = prod;
+        prod = 1; ++j;
+      }
+      (*new_lshape)[j] = l;
+      (*new_rshape)[j] = r;
+      (*new_oshape)[j] = o;
+      ++j;
+    } else {
+      prod *= l;
+    }
+  }
+  if (prod > 1) {
+    (*new_lshape)[j] = (*new_rshape)[j] = (*new_oshape)[j] = prod;
+    ++j;
+  }
+  LOG(INFO) << "o:" << lshape << rshape << oshape;
+  LOG(INFO) << "n:" << *new_lshape << *new_rshape << *new_oshape;
+  if (j != oshape.ndim()) {
+    new_lshape->assign(&(*new_lshape)[0], &(*new_lshape)[j]);
+    new_rshape->assign(&(*new_rshape)[0], &(*new_rshape)[j]);
+    new_oshape->assign(&(*new_oshape)[0], &(*new_oshape)[j]);
+  }
+  LOG(INFO) << "o:" << lshape << rshape << oshape;
+  LOG(INFO) << "n:" << *new_lshape << *new_rshape << *new_oshape;
+  return true;
+}
+
+template<typename xpu, typename OP>
+void BinaryBroadcastCompute(const nnvm::NodeAttrs& attrs,
+                            const OpContext& ctx,
+                            const std::vector<TBlob>& inputs,
+                            const std::vector<OpReqType>& req,
+                            const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mshadow::expr;
+  TShape new_lshape, new_rshape, new_oshape;
+  bool need_bc = BinaryBroadcastShapeCompact(inputs[0].shape_, inputs[1].shape_, outputs[0].shape_,
+                                             &new_lshape, &new_rshape, &new_oshape);
+  if (!need_bc) {
+    BinaryCompute<xpu, OP>(attrs, ctx, inputs, req, outputs);
+  } else {
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    MXNET_NDIM_SWITCH(new_oshape.ndim(), ndim, {
+      MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
```

Comment:
I agree that we want to avoid using DIM_SWICTH and use the original version when possible. Making most things in 3D also potentially make things faster